sdk/storage/AzureStorageBlob/Source/DataStructures/URLSessionTransferManager.swift

```diff
@@ -0,0 +1,389 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//
+// --------------------------------------------------------------------------
+
+import AzureCore
+import CoreData
+
+public final class URLSessionTransferManager: NSObject, TransferManager, URLSessionTaskDelegate {
+    // MARK: Type Aliase
+
+    public typealias TransferManagerType = URLSessionTransferManager
+
+    // MARK: Properties
+
+    public weak var delegate: TransferManagerDelegate?
+
+    public var logger: ClientLogger
+
+    internal lazy var session: URLSession = {
+        let config = URLSessionConfiguration.background(withIdentifier: "com.azuresdk.transfermanager")
+        return URLSession(configuration: config, delegate: self, delegateQueue: nil)
+    }()
+
+    public lazy var reachability: ReachabilityManager? = {
+        var manager = ReachabilityManager()
+        manager?.registerListener { status in
+            switch status {
+            case .unknown, .notReachable:
+                for transfer in self.transfers {
+                    self.pause(transfer: transfer)
+                }
+            case .reachable(.ethernetOrWiFi), .reachable(.wwan):
+                let pausedTransfers = self.transfers.filter { $0.state == .paused }
+                for transfer in pausedTransfers {
+                    self.resume(transfer: transfer)
+                }
+            default:
+                break
+            }
+        }
+        return manager
+    }()
+
+    internal var operationQueue: ResumableOperationQueue
+
+    internal var transfers: [Transfer]
+
+    public var count: Int {
+        return transfers.count
+    }
+
+    public var persistentContainer: NSPersistentContainer? = {
+        guard let bundle = Bundle(identifier: "com.azure.storage.AzureStorageBlob") else { return nil }
+        guard let url = bundle.url(forResource: "AzureStorage", withExtension: "momd") else { return nil }
+        guard let model = NSManagedObjectModel(contentsOf: url) else { return nil }
+        let container = NSPersistentContainer(name: "AzureSDKTransferManager", managedObjectModel: model)
+        container.loadPersistentStores(completionHandler: { _, error in
+            if let error = error as NSError? {
+                // Replace this implementation with code to handle the error appropriately.
```

Comment:
What should we actually do here? Is this a TODO? if so we should mark with a //TODO. It's written like they're instructions for the end user.

Reply:
This is boilerplate XCode gives you when you generate a CoreData project. Realistically, we will need to go through and determine the best way to resolve all instances of "fatalError" in the code (which there are currently many)